sys/kern/vfs_syscalls.c

```diff
@@ -125,6 +125,22 @@ int do_dup2(proc_t *p, int oldfd, int newfd) {
   return 0;
 }
 
+int do_fcntl(proc_t *p, int fd, int cmd, int arg, int *newfdp) {
+  /* TODO: Currently only F_DUPFD command is implemented. */
+  if (cmd != F_DUPFD)
+    return ENOSYS;
+
+  file_t *f;
+  int error;
+
+  if ((error = fdtab_get_file(p->p_fdtable, fd, 0, &f)))
+    return error;
+  if ((error = fdtab_install_file_at_min(p->p_fdtable, f, arg, newfdp)))
```

Comment:
You missed to drop a reference to `f` here in case there was an error.

Reply:
I see... there's similar mistake in other functions in this file. I find current interface of `fdtab_install_file_at` a little bit awkward. I suspect it might be prone to a data race – consider a preemption between `fdtab_get_file` and `fdtab_install_file`.

I'll attempt to quickly improve the interface.